crates/rpcsrv/src/flight/handler.rs

```diff
@@ -101,26 +106,14 @@ impl FlightSessionHandler {
         Ok(Response::new(Box::pin(stream)))
     }
 
-    pub fn new(engine: Arc<Engine>) -> Self {
-        Self {
-            engine,
-            logical_plans: DashMap::new(),
-            sessions: DashMap::new(),
-        }
-    }
-
-    async fn get_or_create_ctx<T>(
-        &self,
-        request: &Request<T>,
-    ) -> Result<Arc<Mutex<Session>>, Status> {
+    async fn get_or_create_ctx<T>(&self, request: &Request<T>) -> Result<SessionRef, Status> {
         let remote = request.remote_addr().unwrap();
 
         let ip = remote.ip().to_string();
         let port = remote.port().to_string();
         let conn_key = ConnKey { ip, port };
-
         if self.sessions.contains_key(&conn_key) {
-            let sess = self.sessions.get(&conn_key).unwrap().clone();
+            let sess = self.sessions.get(&conn_key).unwrap();
```

Comment:
is there a way to have the connections/sessions keep track of active callers/users/etc and have them drop themselves if they've been idle for a certain period of time. (e.g. have something in the destructor)

Reply:
not that I'm aware of. 

If you are aware of a solution that automatically manages the lifecycles based on the connection timeout, I think that would be much better and would happily do that over this cache based approach.

Reply:
I think 
- there's this approach (or some variant therein), with a single worker that goes through and cleans up stale connections;
- you could write a version where the worker thread is triggered by a condition variable and some other event (the drop method on the connection object could trigger a cleanup pass);
- you store last-used in the map along with an Arc wrapped connection, and have a monitor task for each connection (maybe isolated to a simple runtime if it matters) that either sleeps till the ttl or drops the reference or if there are no other holders of the session.